In [1]:
from pathlib import Path

import pandas as pd
import numpy as np

import matplotlib.pyplot as plt
import seaborn as sns

from pathlib import Path


In [3]:
print(Path.cwd())

c:\Users\HP\Desktop\Projects\airshift\notebooks


In [6]:
DATA_DIR = Path("../data/raw")

print("Data directory:")
print(DATA_DIR.resolve())

print("\nDoes it exist?")
print(DATA_DIR.exists())

print("\nContents:")
for item in DATA_DIR.iterdir():
    print(item.name)

Data directory:
C:\Users\HP\Desktop\Projects\airshift\data\raw

Does it exist?
True

Contents:
PRSA_Data_Aotizhongxin_20130301-20170228.csv
PRSA_Data_Changping_20130301-20170228.csv
PRSA_Data_Dingling_20130301-20170228.csv
PRSA_Data_Dongsi_20130301-20170228.csv
PRSA_Data_Guanyuan_20130301-20170228.csv
PRSA_Data_Gucheng_20130301-20170228.csv
PRSA_Data_Huairou_20130301-20170228.csv
PRSA_Data_Nongzhanguan_20130301-20170228.csv
PRSA_Data_Shunyi_20130301-20170228.csv
PRSA_Data_Tiantan_20130301-20170228.csv
PRSA_Data_Wanliu_20130301-20170228.csv
PRSA_Data_Wanshouxigong_20130301-20170228.csv


In [7]:
files = sorted(DATA_DIR.glob("*.csv"))

print(f"Number of files: {len(files)}")

Number of files: 12


# Profiling

## 1. Profiling Aotizhongxin Station file to understand the file format

In [8]:
sample_file = files[0]

sample_df = pd.read_csv(sample_file)

print("File:", sample_file.name)
print("Shape:", sample_df.shape)

File: PRSA_Data_Aotizhongxin_20130301-20170228.csv
Shape: (35064, 18)


In [9]:
sample_df.head()

,No,year,month,day,hour,PM2.5,PM10,SO2,NO2,CO,O3,TEMP,PRES,DEWP,RAIN,wd,WSPM,station
0,1,2013,3,1,0,4.0,4.0,4.0,7.0,300.0,77.0,-0.7,1023.0,-18.8,0.0,NNW,4.4,Aotizhongxin
1,2,2013,3,1,1,8.0,8.0,4.0,7.0,300.0,77.0,-1.1,1023.2,-18.2,0.0,N,4.7,Aotizhongxin
2,3,2013,3,1,2,7.0,7.0,5.0,10.0,300.0,73.0,-1.1,1023.5,-18.2,0.0,NNW,5.6,Aotizhongxin
3,4,2013,3,1,3,6.0,6.0,11.0,11.0,300.0,72.0,-1.4,1024.5,-19.4,0.0,NW,3.1,Aotizhongxin
4,5,2013,3,1,4,3.0,3.0,12.0,12.0,300.0,72.0,-2.0,1025.2,-19.5,0.0,N,2.0,Aotizhongxin


In [10]:
sample_df.tail()

,No,year,month,day,hour,PM2.5,PM10,SO2,NO2,CO,O3,TEMP,PRES,DEWP,RAIN,wd,WSPM,station
35059,35060,2017,2,28,19,12.0,29.0,5.0,35.0,400.0,95.0,12.5,1013.5,-16.2,0.0,NW,2.4,Aotizhongxin
35060,35061,2017,2,28,20,13.0,37.0,7.0,45.0,500.0,81.0,11.6,1013.6,-15.1,0.0,WNW,0.9,Aotizhongxin
35061,35062,2017,2,28,21,16.0,37.0,10.0,66.0,700.0,58.0,10.8,1014.2,-13.3,0.0,NW,1.1,Aotizhongxin
35062,35063,2017,2,28,22,21.0,44.0,12.0,87.0,700.0,35.0,10.5,1014.4,-12.9,0.0,NNW,1.2,Aotizhongxin
35063,35064,2017,2,28,23,19.0,31.0,10.0,79.0,600.0,42.0,8.6,1014.1,-15.9,0.0,NNE,1.3,Aotizhongxin


In [11]:
sample_df.info()

<class 'pandas.DataFrame'>
RangeIndex: 35064 entries, 0 to 35063
Data columns (total 18 columns):
 #   Column   Non-Null Count  Dtype  
---  ------   --------------  -----  
 0   No       35064 non-null  int64  
 1   year     35064 non-null  int64  
 2   month    35064 non-null  int64  
 3   day      35064 non-null  int64  
 4   hour     35064 non-null  int64  
 5   PM2.5    34139 non-null  float64
 6   PM10     34346 non-null  float64
 7   SO2      34129 non-null  float64
 8   NO2      34041 non-null  float64
 9   CO       33288 non-null  float64
 10  O3       33345 non-null  float64
 11  TEMP     35044 non-null  float64
 12  PRES     35044 non-null  float64
 13  DEWP     35044 non-null  float64
 14  RAIN     35044 non-null  float64
 15  wd       34983 non-null  str    
 16  WSPM     35050 non-null  float64
 17  station  35064 non-null  str    
dtypes: float64(11), int64(5), str(2)
memory usage: 4.8 MB


In [12]:
sample_df.describe(include="all").T

,count,unique,top,freq,mean,std,min,25%,50%,75%,max
No,35064.0,NaN,NaN,NaN,17532.5,10122.249256,1.0,8766.75,17532.5,26298.25,35064.0
year,35064.0,NaN,NaN,NaN,2014.66256,1.177213,2013.0,2014.0,2015.0,2016.0,2017.0
month,35064.0,NaN,NaN,NaN,6.52293,3.448752,1.0,4.0,7.0,10.0,12.0
day,35064.0,NaN,NaN,NaN,15.729637,8.800218,1.0,8.0,16.0,23.0,31.0
hour,35064.0,NaN,NaN,NaN,11.5,6.922285,0.0,5.75,11.5,17.25,23.0
PM2.5,34139.0,NaN,NaN,NaN,82.773611,82.135694,3.0,22.0,58.0,114.0,898.0
PM10,34346.0,NaN,NaN,NaN,110.060391,95.223005,2.0,38.0,87.0,155.0,984.0
SO2,34129.0,NaN,NaN,NaN,17.375901,22.823017,0.2856,3.0,9.0,21.0,341.0
NO2,34041.0,NaN,NaN,NaN,59.305833,37.1162,2.0,30.0,53.0,82.0,290.0
CO,33288.0,NaN,NaN,NaN,1262.945145,1221.436236,100.0,500.0,900.0,1500.0,10000.0


In [13]:
sample_df.columns.tolist()

['No',
 'year',
 'month',
 'day',
 'hour',
 'PM2.5',
 'PM10',
 'SO2',
 'NO2',
 'CO',
 'O3',
 'TEMP',
 'PRES',
 'DEWP',
 'RAIN',
 'wd',
 'WSPM',
 'station']

### Missing Value

In [14]:
missing = sample_df.isnull().sum()

missing = missing[missing > 0].sort_values(ascending=False)

missing

CO       1776
O3       1719
NO2      1023
SO2       935
PM2.5     925
PM10      718
wd         81
TEMP       20
PRES       20
DEWP       20
RAIN       20
WSPM       14
dtype: int64

In [15]:
missing_percentage = (
    sample_df.isnull().mean() * 100
).sort_values(ascending=False)

missing_percentage[missing_percentage > 0]

CO       5.065024
O3       4.902464
NO2      2.917522
SO2      2.666553
PM2.5    2.638033
PM10     2.047684
wd       0.231006
RAIN     0.057039
DEWP     0.057039
PRES     0.057039
TEMP     0.057039
WSPM     0.039927
dtype: float64

### Duplication

In [16]:
print("Duplicate rows:", sample_df.duplicated().sum())

Duplicate rows: 0


## 🔍 Initial Data Profiling for Aotizhongxin Station file

| Aspect         | Finding                                              |
| -------------- | ---------------------------------------------------- |
| Observations   | 35,064 hourly records                                |
| Features       | 18                                                   |
| Period         | March 2013 – February 2017                           |
| Data types     | Air pollution, meteorological, and temporal features |
| Missing values | Mainly `CO` (5.07%) and `O3` (4.90%)                 |
| Duplicates     | None detected                                        |
| PM2.5 range    | 3 – 898                                              |
| Data cleaning  | Not performed at this stage                          |


## 2. Cross-Station Profiling

After the initial profiling of one station, we compare all 12 stations to assess their structure, size, missing values, duplicates, and time coverage before data cleaning.


In [17]:
profiling_summary = []

for file in files:
    df = pd.read_csv(file)

    profiling_summary.append({
        "station": df["station"].iloc[0],
        "rows": df.shape[0],
        "columns": df.shape[1],
        "missing_values": df.isnull().sum().sum(),
        "missing_percentage": df.isnull().mean().mean() * 100,
        "duplicate_rows": df.duplicated().sum(),
        "start_year": df["year"].min(),
        "end_year": df["year"].max()
    })

profiling_df = pd.DataFrame(profiling_summary)

profiling_df

,station,rows,columns,missing_values,missing_percentage,duplicate_rows,start_year,end_year
0,Aotizhongxin,35064,18,7271,1.152020,0,2013,2017
1,Changping,35064,18,5166,0.818503,0,2013,2017
2,Dingling,35064,18,7015,1.111460,0,2013,2017
3,Dongsi,35064,18,7600,1.204147,0,2013,2017
4,Guanyuan,35064,18,5279,0.836407,0,2013,2017
5,Gucheng,35064,18,4728,0.749106,0,2013,2017
6,Huairou,35064,18,7485,1.185927,0,2013,2017
7,Nongzhanguan,35064,18,4090,0.648021,0,2013,2017
8,Shunyi,35064,18,8523,1.350388,0,2013,2017
9,Tiantan,35064,18,5277,0.836090,0,2013,2017


In [18]:
profiling_df.sort_values(
    by="missing_percentage",
    ascending=False
)

,station,rows,columns,missing_values,missing_percentage,duplicate_rows,start_year,end_year
8,Shunyi,35064,18,8523,1.350388,0,2013,2017
3,Dongsi,35064,18,7600,1.204147,0,2013,2017
6,Huairou,35064,18,7485,1.185927,0,2013,2017
0,Aotizhongxin,35064,18,7271,1.152020,0,2013,2017
2,Dingling,35064,18,7015,1.111460,0,2013,2017
10,Wanliu,35064,18,6447,1.021466,0,2013,2017
4,Guanyuan,35064,18,5279,0.836407,0,2013,2017
9,Tiantan,35064,18,5277,0.836090,0,2013,2017
1,Changping,35064,18,5166,0.818503,0,2013,2017
11,Wanshouxigong,35064,18,5146,0.815334,0,2013,2017


In [19]:
reference_columns = set(pd.read_csv(files[0]).columns)

for file in files:
    df = pd.read_csv(file)
    
    if set(df.columns) == reference_columns:
        print(f"{file.name}: Columns match")
    else:
        print(f"{file.name}: Columns differ")

PRSA_Data_Aotizhongxin_20130301-20170228.csv: Columns match
PRSA_Data_Changping_20130301-20170228.csv: Columns match
PRSA_Data_Dingling_20130301-20170228.csv: Columns match
PRSA_Data_Dongsi_20130301-20170228.csv: Columns match
PRSA_Data_Guanyuan_20130301-20170228.csv: Columns match
PRSA_Data_Gucheng_20130301-20170228.csv: Columns match
PRSA_Data_Huairou_20130301-20170228.csv: Columns match
PRSA_Data_Nongzhanguan_20130301-20170228.csv: Columns match
PRSA_Data_Shunyi_20130301-20170228.csv: Columns match
PRSA_Data_Tiantan_20130301-20170228.csv: Columns match
PRSA_Data_Wanliu_20130301-20170228.csv: Columns match
PRSA_Data_Wanshouxigong_20130301-20170228.csv: Columns match


### Key Findings

| Finding              | Result                               |
| -------------------- | ------------------------------------ |
| Dataset size         | 35,064 rows × 18 columns per station |
| Time coverage        | 2013–2017 for all stations           |
| Column structure     | Consistent across all 12 stations    |
| Duplicate rows       | None detected                        |
| Missing values       | 0.65%–1.35% across stations          |
| Highest missing rate | Shunyi (1.35%)                       |
| Lowest missing rate  | Nongzhanguan (0.65%)                 |


## 3. Temporal Profiling

Since AirShift is based on hourly time-series data, we first verify the temporal consistency of all monitoring stations before handling missing values.

### 3.1 Timestamp Consistency

In [20]:
temporal_summary = []

for file in files:
    df = pd.read_csv(file)

    df["datetime"] = pd.to_datetime(
        df[["year", "month", "day", "hour"]]
    )

    temporal_summary.append({
        "station": df["station"].iloc[0],
        "start": df["datetime"].min(),
        "end": df["datetime"].max(),
        "rows": len(df),
        "unique_timestamps": df["datetime"].nunique(),
        "duplicate_timestamps": df["datetime"].duplicated().sum(),
        "chronologically_sorted": df["datetime"].is_monotonic_increasing
    })

temporal_df = pd.DataFrame(temporal_summary)

temporal_df

,station,start,end,rows,unique_timestamps,duplicate_timestamps,chronologically_sorted
0,Aotizhongxin,2013-03-01,2017-02-28 23:00:00,35064,35064,0,True
1,Changping,2013-03-01,2017-02-28 23:00:00,35064,35064,0,True
2,Dingling,2013-03-01,2017-02-28 23:00:00,35064,35064,0,True
3,Dongsi,2013-03-01,2017-02-28 23:00:00,35064,35064,0,True
4,Guanyuan,2013-03-01,2017-02-28 23:00:00,35064,35064,0,True
5,Gucheng,2013-03-01,2017-02-28 23:00:00,35064,35064,0,True
6,Huairou,2013-03-01,2017-02-28 23:00:00,35064,35064,0,True
7,Nongzhanguan,2013-03-01,2017-02-28 23:00:00,35064,35064,0,True
8,Shunyi,2013-03-01,2017-02-28 23:00:00,35064,35064,0,True
9,Tiantan,2013-03-01,2017-02-28 23:00:00,35064,35064,0,True


### Temporal Profiling Findings

All 12 stations have consistent temporal coverage from **March 2013 to February 2017** with 35,064 unique hourly timestamps each. No duplicate timestamps were detected, and all datasets are chronologically ordered.

The next step is to check for **missing hourly timestamps (time gaps)**.


### 3.2 Time Gap Analysis

We check for missing hourly timestamps to ensure the time series is continuous across all monitoring stations.

In [21]:
gap_summary = []

for file in files:
    df = pd.read_csv(file)

    df["datetime"] = pd.to_datetime(
        df[["year", "month", "day", "hour"]]
    )

    time_diff = df["datetime"].diff()

    gaps = (time_diff > pd.Timedelta(hours=1)).sum()

    gap_summary.append({
        "station": df["station"].iloc[0],
        "time_gaps": gaps,
        "largest_gap": time_diff.max()
    })

gap_df = pd.DataFrame(gap_summary)

gap_df

,station,time_gaps,largest_gap
0,Aotizhongxin,0,0 days 01:00:00
1,Changping,0,0 days 01:00:00
2,Dingling,0,0 days 01:00:00
3,Dongsi,0,0 days 01:00:00
4,Guanyuan,0,0 days 01:00:00
5,Gucheng,0,0 days 01:00:00
6,Huairou,0,0 days 01:00:00
7,Nongzhanguan,0,0 days 01:00:00
8,Shunyi,0,0 days 01:00:00
9,Tiantan,0,0 days 01:00:00


### Temporal Profiling Summary

| Check                | Result              | Conclusion                     |
| -------------------- | ------------------- | ------------------------------ |
| Time coverage        | Mar 2013 – Feb 2017 | Consistent across all stations |
| Records per station  | 35,064              | Equal across all stations      |
| Unique timestamps    | 35,064              | No repeated timestamps         |
| Duplicate timestamps | 0                   | No duplicates detected         |
| Chronological order  | Yes                 | Data is correctly ordered      |
| Time gaps            | 0                   | Continuous hourly timeline     |
| Maximum interval     | 1 hour              | Expected hourly frequency      |